python-lib/dku_webapp/utils.py

```diff
@@ -0,0 +1,182 @@
+    target_variable = model_accessor.get_target_variable()
+    test_df.dropna(subset=[sensitive_column, target_variable], how='any', inplace=True)
+
+    y_true = test_df.loc[:, target_variable]
+    pred_df = model_accessor.predict(test_df)
+    y_pred = pred_df.loc[:, DkuWebappConstants.PREDICTION]
+
+
+    try: # check whether or not the column can be casted to int
+        if np.array_equal(test_df[sensitive_column], test_df[sensitive_column].astype(int)):
+            test_df[sensitive_column] = test_df[sensitive_column].astype(int)
+        if test_df[sensitive_column].dtypes == int:
+            reference_group = int(reference_group)
+        if test_df[sensitive_column].dtypes == float:
+            reference_group = float(reference_group)
+    except:
+        pass
+
+    sensitive_feature_values = test_df[sensitive_column]
+    model_report = ModelFairnessMetricReport(y_true, y_pred, sensitive_feature_values, advantageous_outcome)
+    population_names = sensitive_feature_values.unique()
+
+    metric_dct = {}
+    disparity_dct = {}
+    for metric_func in ModelFairnessMetric.get_available_metric_functions():
+        metric_summary = model_report.compute_metric_per_group(metric_function=metric_func)
+        metric_dct[metric_func.__name__] = metric_summary.get(DkuFairnessConstants.BY_GROUP)
+        metric_diff = model_report.compute_group_difference_from_summary(metric_summary, reference_group=reference_group)
+        v = np.array(list(metric_diff.get(DkuFairnessConstants.BY_GROUP).values())).reshape(1, -1).squeeze()
+        v_without_nan = [x for x in v if not np.isnan(x)]
+        if len(v_without_nan) > 0:
+            max_disparity = max(v_without_nan, key=abs)
+            disparity_dct[metric_func.__name__] = max_disparity
+        else:
+            disparity_dct[metric_func.__name__] = 'N/A' # for display purpose
+
+    populations = []
+    for name in population_names:
+        dct = {
+            DkuWebappConstants.NAME: name,
+            DkuWebappConstants.SIZE: len(test_df[test_df[sensitive_column] == name])
+        }
+        for m, v in metric_dct.items():
+            # the following strings are used only here, too lazy to turn them into constant variables
+            if m == 'demographic_parity':
+                dct['positive_rate'] = v[name]
+            if m == 'equalized_odds':
+                dct['true_positive_rate'], dct['false_positive_rate'] = v[name]
+            if m == 'predictive_rate_parity':
+                dct['positive_predictive_value'] = v[name]
+
+        # make sure that NaN is replaced by a string (a dot here), for display purpose
+        for k, v in dct.items():
+            if not isinstance(v, str) and np.isnan(v):
+                dct[k] = '.'
+        populations.append(dct)
+
+    label_list = model_report.get_label_list()
+
+    return sorted(populations, key=lambda population: population[DkuWebappConstants.SIZE], reverse=True), disparity_dct, label_list
```

Comment:
Improve readability and define a variable on sorted(populations, key=lambda population: population[DkuWebappConstants.SIZE], reverse=True) before return

Reply:
Done in e4c1e70